core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -2790,7 +2790,10 @@ object SparkContext extends Logging {
             s" = ${taskReq.amount}")
         }
         // Compare and update the max slots each executor can provide.
-        val resourceNumSlots = execAmount / taskReq.amount
+        // If the configured amount per task was < 1.0, a task is subdividing
+        // executor resources. If the amount per task was > 1.0, the task wants
+        // multiple executor resources.
+        val resourceNumSlots = Math.floor(execAmount * taskReq.numParts / taskReq.amount).toInt
```

Comment:
IIUC, either `numParts` or `amount` should be 1 here? Shall we introduce a var or function to `ResourceRequirement` to represent `amount / numParts` to simplify the code?

Reply:
I think that you could go either way. If you feel strongly that this is confusing, I can try to change the case class. I like the formula here because it shows you explicitly how slots are computed.